### PR TITLE
swr: improve text cursor placement by mouse and arrow keys

### PR DESCRIPTION
**Pull request by FloVanGH**

### CODE REVIEW

File: internal/core/textlayout.rs [hunk 1]
```diff
@@ -255,14 +255,19 @@ impl<'a, Font: AbstractFont> TextParagraphLayout<'a, Font> {
 
     /// Returns the bytes offset for the given position
     pub fn byte_offset_for_position(&self, (pos_x, pos_y): (Font::Length, Font::Length)) -> usize {
-        let byte_offset = 0;
+        let mut byte_offset = 0;
         let two = Font::LengthPrimitive::one() + Font::LengthPrimitive::one();
 
         match self.layout_lines(|glyphs, _, line_y, line| {
             if pos_y >= line_y + self.layout.font.height() {
+                byte_offset = line.byte_range.end;
```
Comment (tronical) [inline internal/core/textlayout.rs:263]: Is this for the case of clicking below the last line?

**Review by tronical — CHANGES_REQUESTED** [review]
Looks good. Please add unit test coverage for this though.

File: internal/core/textlayout.rs [hunk 2]
```diff
@@ -255,14 +255,19 @@ impl<'a, Font: AbstractFont> TextParagraphLayout<'a, Font> {
...
                 return core::ops::ControlFlow::Continue(());
             }
 
+            if line.is_empty() {
+                return core::ops::ControlFlow::Break(line.byte_range.start);
```
[the remark this answers is not in the document]
Comment (FloVanGH) [reply internal/core/textlayout.rs:268]: Sure.

File: internal/core/textlayout.rs (hunk 1 above)
Comment (tronical) [reply internal/core/textlayout.rs:263]: Ah right. Okay, that should be easy to test by extending one of the existing tests :)

File: internal/core/textlayout.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (FloVanGH) [reply internal/core/textlayout.rs:268]: done

File: internal/core/textlayout.rs [hunk 3]
```diff
@@ -532,6 +537,26 @@ fn test_cursor_position_with_newline() {
...
+        wrap: TextWrap::WordWrap,
+        overflow: TextOverflow::Clip,
+        single_line: false,
+    };
+
+    assert_eq!(paragraph.byte_offset_for_position((0., 10.)), 6);
+}
+
 #[test]
 fn test_byte_offset() {
```
Comment (tronical) [inline internal/core/textlayout.rs:561]: Please extend this test to verify that `paragraph.byte_offset_for_position` returns the last byte offset when the y position is below the last line. I believe that part isn't covered yet, or did I miss that?

**Review by tronical — APPROVED** [review]
Thanks :)

**Pull request merged**